activemq-unit-tests/src/test/java/org/apache/activemq/jms2/ActiveMQJMS2MessageTypesTest.java

```diff
@@ -0,0 +1,582 @@
+                        foundCount++;
+                    }
+                }
+            }
+            assertEquals(Integer.valueOf(2), Integer.valueOf(foundCount));
+            jmsConsumer.close();
+        } catch (Exception e) {
+            fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testMessageDeliveryModeInvalid() {
+
+        try(JMSContext jmsContext = activemqConnectionFactory.createContext("admin", "admin", Session.AUTO_ACKNOWLEDGE)) {
+            assertNotNull(jmsContext);
+            Destination destination = ActiveMQJMS2TestSupport.generateDestination(jmsContext, destinationType, methodNameDestinationName);
+            assertNotNull(destination);
+            jmsContext.start();
+
+            Message message = ActiveMQJMS2TestSupport.generateMessage(jmsContext, messageType, messagePayload);
+            boolean caught = false;
+            try {
+                ActiveMQJMS2TestSupport.sendMessage(jmsContext, destination, message, null, 99, null, null, null, null, null, null, null);
+                fail("IlegalStateRuntimeException expected");
+            } catch (IllegalStateRuntimeException e) {
+                assertEquals("unknown delivery mode: 99", e.getMessage());
+                caught = true;
+            }
+            assertTrue(caught);
+        } catch (Exception e) {
+            fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testMessagePriority() {
+
+        try(JMSContext jmsContext = activemqConnectionFactory.createContext("admin", "admin", Session.AUTO_ACKNOWLEDGE)) {
+            assertNotNull(jmsContext);
+            Destination destination = ActiveMQJMS2TestSupport.generateDestination(jmsContext, destinationType, methodNameDestinationName);
+            assertNotNull(destination);
+            JMSConsumer jmsConsumer = jmsContext.createConsumer(destination);
+            assertNotNull(jmsConsumer);
+            jmsContext.start();
+
+            Message message = ActiveMQJMS2TestSupport.generateMessage(jmsContext, messageType, messagePayload);
+
+            List<String> sentMessageIds = new LinkedList<>();
+            for(int priority=0; priority<10; priority++) {
+                sentMessageIds.add(ActiveMQJMS2TestSupport.sendMessage(jmsContext, destination, message, null, null, null, null, null, null, null, priority, null));
+            }
+
+            Message recvMessage = null;
+            List<Message> recvMessages = new LinkedList<>();
+            int loopCount = 0;
+            int maxLoops = 50;
+            boolean done = false;
+            do {
+                recvMessage = jmsConsumer.receive(500l);
+                if(recvMessage != null) {
+                    recvMessages.add(recvMessage);
+                }
+
+                if(recvMessages.size() == 10) {
+                    done = true;
+                }
+                loopCount++;
+            } while (loopCount <= maxLoops && !done);
```

Comment:
I agree that this test pattern needs to be improved. I had brought this up before when working tests for 5.17.x so I think now would be a good time to rewrite this as it's re-used in several tests.

Reply:
I will add at the very least you can at least turn it into a helper method and re-use the code that way it can be easily refined or tuned later vs where it's copy and pasted all over the place now (that was my main complaint from 5.17.x where the same do/while was copy pasted over and over so if you wanted to change the design it would be awful to fix)

Reply:
Updated